source/slang/slang-parser.cpp

```diff
@@ -3314,24 +3342,66 @@ namespace Slang
             // expression statement, and we need to figure out which.
             //
             // We'll solve this with backtracking for now.
+            //
+            // TODO: This should not require backtracking at all.
 
             TokenReader::ParsingCursor startPos = tokenReader.getCursor();
 
             // Try to parse a type (knowing that the type grammar is
             // a subset of the expression grammar, and so this should
             // always succeed).
+            //
+            // HACK: The type grammar that `ParseType` supports is *not*
+            // a subset of the expression grammar because it includes
+            // type specififers like `struct` and `enum` declarations
+            // which should always be the start of a declaration.
+            //
+            // TODO: Before launching into this attempt to parse a type,
+            // this logic should really be looking up the `SyntaxDecl`,
+            // if any, assocaited with the identifier. If a piece of
+            // syntax is discovered, then it should dictate the next
+            // steps of parsing, and only in the case where the lookahead
+            // isn't a keyword should we fall back to the approach
+            // here.
+            //
             RefPtr<Expr> type = ParseType();
+
             // We don't actually care about the type, though, so
             // don't retain it
+            //
+            // TODO: There is no reason to throw away the work we
+            // did parsing the `type` expression. Once we disambiguate
+            // what to do, we should be able to use the expression
+            // we already parsed as a starting point for whatever we
+            // parse next. E.g., if we have `A.b` and the lookahead is `+`
+            // then we can use `A.b` as the left-hand-side expression
+            // when starting to parse an infix expression.
+            //
             type = nullptr;
 
+            // TODO: If we decide to intermix parsing of statement bodies
+            // with semantic checking (by delaying the parsing of bodies
+            // until semantic context is available), then we could look
+            // at the *type* of `type` to disambiguate what to do next,
+            // which might result in a nice simplification (at the cost
+            // of definitely making the grammar context-dependent).
+
             // If the next token after we parsed a type looks like
             // we are going to declare a variable, then lets guess
             // that this is a declaration.
             //
             // TODO(tfoley): this wouldn't be robust for more
             // general kinds of declarators (notably pointer declarators),
             // so we'll need to be careful about this.
+            //
+            // If the lookahead token is `*`, then we have to decide
```

Comment:
I guess you could operator overload * and do something that had side effects, but probably not a significant issue for now even if true.

Reply:
Correct, an operator overload could violate the assumptions, but it would still be really odd to write `a * b;` as a statement, to the point that I don't think the language would be unreasonable in banning it.

It might still be surprising if you are refactoring/editing a shader and you end up with a stray subexpression on its own line and the compiler gives you an error like "hey, `a` is not a type!" and/or "hey, you already *have* a variable called `b` in this function!".

I suspect that the best long-term answer (if we continue to support C-style declaration syntax, which I don't think we have any choice about) will be to make our parsing of `{}`-enclosed bodies be context-sensitive, so that we intermix semantic checking and parsing just like a C/C++ compiler would. That would let us more easily parse tricky cases of the C/C++ syntax without error, and even in cases where we don't want to support some aspects of that syntax it would make our error messages more robust and tailored if we have the additional context.

The global scope would still be parsed in a context-free fashion to support out-of-order declarations, but most of the ambiguity is gone there because of the lack of support for statements/expressions. The one tricky bit that would come up at the global scope would be the "most vexing parse" sorts of situations which all use C++ variable declaration+construction syntax:

```c++
Foo bar(Baz); // a function that takes an (unnamed) parameter of type `Baz`
Foo bar(baz); // a variable that is being initialized/constructed with the argument value `baz`

// Ambiguous:
// * A function that takes an unnamed parameter of type `Baz`?
// * A variable that is initialized/constructed with a default-constructed `Baz` value?
// (the latter choice is arbitrarily selected)
//
Foo bar(Baz());
```

If we ban C++-style variable declarations with constructor arguments at global scope, this would be a non-issue, but unfortunately the DXR extensions technically introduced a few global types used for DXIL collection/linkage meta-data stuff that are designed to be used with constructor syntax.

We may need to define a plan for that stuff sooner or later, but the fact that C++ variable+constructor syntax is a complete mess makes me disinclined to spend too much effort on trying to support it.